### ISSUE TITLE
Dataloader doesn't batch?

### ISSUE THREAD

**Issue by mikepuglisi** [Narkoleptika/apollo-api-boilerplate#1]
I'm trying to get a better understanding of using Dataloader, particularly with Sequelize relationships. Your boilerplate looks very close to solving the problem I'm having, but I'm not sure it is batching results as it should be. Perhaps it is my lack of understanding of how dataloader *should* work.

For example, given the following query:

```
{
  getPosts {
    title
    content
    comments {
      content
    }
  }
}
```

I expect dataloader to batch the comments query with an IN statement so there is only one query for comments. What actually happens is the following:

```
Executing (default): SELECT `id`, `title`, `content`, `createdAt`, `updatedAt` FROM `posts` AS `post` ORDER BY `post`.`createdAt` DESC;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 1;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 2;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 3;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 4;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 5;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 6;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 7;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 8;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 9;
Executing (default): SELECT `id`, `content`, `createdAt`, `updatedAt`, `postId` FROM `comments` AS `comment` WHERE `comment`.`postId` = 10;
```

Shouldn't this be something like this?

WHERE `comment`.`postId` IN (1,2,3,4,5,6,7,8,9,10);

I though that was the whole point of batching with dataloader.

**Comment by Narkoleptika** [comment]
@mikepuglisi 

Thanks, your issue illuminates an optimization opportunity.

You are correct that it would be better to run a where in query, but that's not Dataloader's responsibility.

You'll find that if you run the following contrived query the above sql queries shouldn't run a second time.
```gql
{
  getPosts {
    title
    content
    comments {
      content
      post {
        title
        content
        comments {
          content
        }
      }
    }
  }
}
```

Dataloader is just going to batch some ids, run the resolver that's defined and then cache the results. This is especially apparent for a many to many relationship where multiple records might try to query for the same related record which would cause duplicate sql queries without Dataloader. 

If you look here: https://github.com/Narkoleptika/apollo-api-boilerplate/blob/master/dataloaders/comment.js#L6

You'll see that I'm taking those batched ids, mapping through and returning a query for each.

To get what you want (and what I should update the boilerplate to do), the code linked above should be changed to something like the following (haven't tested)

```js
const DataLoader = require('dataloader')
const {comment, Sequelize: {Op}} = require('../db/models')

module.exports = () => ({
    postCommentLoader: new DataLoader(async postIds =>
        await comment.findAll({where: {postId: {[Op.in]: postIds}}})
    )
})
```

I'll update the boilerplate with this change when I get a chance to test.

**Comment by mikepuglisi** [comment]
Thanks! Ultimately I'm building up to a use case that has many to many relationship with a junction table containing it's own relevant information as well. Trying to wrap my head around all of this. 

As for updating postCommentLoader to use an IN clause, I'm trying to implement that as well. The solution you provided above results in the following error. I suspect it is because the data isn't shaped they way it needs to be shaped...?

![image](https://user-images.githubusercontent.com/37956042/48504273-56e98500-e812-11e8-92c2-26ba13e08ab5.png)

I'm trying to solve it by mapping the results. Not sure though.

**Comment by Narkoleptika** [comment]
@mikepuglisi 

Ah this may be why I didn't use a where in initially. 

The problem is that the resolver function was passed an array for postIds of `[1,2,3,4,5,6,7,8,9,10]` and presumably each post has more than one comment so it returned an array of more than 10 comments. Dataloader expects a promise that resolves to an array of the same length as the one that was passed in for the return value.

In order to use a where in, you'll have to do some additional processing to make sure that the return value looks like this given the input of `[1,2,3,4,5,6,7,8,9,10]`

```js
[
    [comments for post 1],
    [comments for post 2],
    [comments for post 3],
    [comments for post 4],
    [comments for post 5],
    [comments for post 6],
    [comments for post 7],
    [comments for post 8],
    [comments for post 9],
    [comments for post 10]
]
```

**Comment by mikepuglisi** [comment]
Yep. I just got it done.

```
const DataLoader = require('dataloader')
const {comment} = require('../db/models')

module.exports = () => ({
    postCommentLoader: new DataLoader(async postIds => await comment.findAll({ where: {postId: postIds}})
      .then(rows => postIds.map(postId => rows.filter(x => x.postId === postId)))
    )
})

```

Note that you don't need the [Op.in], since passing an array handles that automatically. I suspect this operation is better than doing n+1 queries. Not sure.

What is your take on the dataloader-sequelize library? I'm wondering if I should explore something like that rather than implement dataloader manually. I see you are doing it manually, just wondering if it was a conscious decision to not go with something like dataloader-sequelize.

**Comment by Narkoleptika** [comment]
@mikepuglisi 

Thanks for the excellent tip about Op.in!

Nope, never heard about it until now.

Unless there is an extremely compelling reason to switch, I'd probably opt to continue using it manually.

Especially since one of GraphQL's selling points is the ability to use multiple different data sources under the hood.

So it's entirely likely that someone would want to, for example, include some rest api, or mongo as a data source. In which case, it would be better to have a more general dataloader implementation that could be used with sequelize and the rest api/mongo.

Closed by 975f401fbf99edce82a0beace3435ddc95d45349
